### PR TITLE
Implement referrer policy for WorkerGlobalScope

### LINKED ISSUE #11863: Implement referrer policy for WorkerGlobalScope
See spec at https://w3c.github.io/webappsec-referrer-policy/. All servo referrer policy implementation thus far is just for documents. See #10311 for reference.

### PR DESCRIPTION
**Pull request by aravind-pg**
Adds a new `Option<ReferrerPolicy>` field to WorkerGlobalScope and sets it appropriately by parsing the header when setting up a ServiceWorkerGlobalScope or a DedicatedWorkerGlobalScope. 

Also adds a new referrer policy accessor to GlobalRef and uses it in XMLHttpRequest's constructor.
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #11863 
- [X] There are tests for these changes

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/12473)

### CODE REVIEW

**Comment by aravind-pg** [comment]
Only one WPT test was fixed by this PR, so I'm not sure I didn't do something wrong.

r? @jdm cc @rebstar6

**Comment by jdm** [comment]
I believe all of the worker test infrastructure in referrer-policy is disabled. I haven't looked into why that's the case yet.

**Comment by aravind-pg** [comment]
Ah I vaguely recall reading about this somewhere but can't seem to find it now. Could you point me to where exactly they're being disabled? I wasn't able to find anything in my look-around.

**Comment by jdm** [comment]
It's the worker-request thing in spec.src.json. Maybe it's excluded, or just not included.

**Comment by aravind-pg** [comment]
Updated to re-enable the worker-requests tests, a bunch of which did pass :)

**Comment by aravind-pg** [comment]
Rebased.

File: components/script/dom/xmlhttprequest.rs [hunk 1]
```diff
@@ -153,13 +153,8 @@ pub struct XMLHttpRequest {
...
 impl XMLHttpRequest {
     fn new_inherited(global: GlobalRef) -> XMLHttpRequest {
-        //TODO - update this when referrer policy implemented for workers
-        let (referrer_url, referrer_policy) = if let GlobalRef::Window(window) = global {
-            let document = window.Document();
-            (Some(document.url().clone()), document.get_referrer_policy())
-        } else {
-            (None, None)
-        };
+        let referrer_url = Some(global.get_url());
```
Comment (aravind-pg) [inline components/script/dom/xmlhttprequest.rs]: Note that now the field `referrer_url` is always `Some`, so I don't think it needs to stay an option any more?

**Comment by aravind-pg** [comment]
Rebased.

**Comment by jdm** [comment]
I'll review this on Monday. Sorry for the wait!

**Comment by aravind-pg** [comment]
No problem at all.

File: components/script/dom/xmlhttprequest.rs (hunk 1 above)
Comment (jdm) [inline components/script/dom/xmlhttprequest.rs]: Seems reasonable!

**Comment by jdm** [comment]
The fact that we're failing tests like `http-http/worker-request/insecure-protocol.keep-origin-redirect.http.html.ini` make me suspect that either the tests aren't working right or our implementation isn't right. The changes in this PR look like what I expected to see, but we should track down the source of the test failure before moving ahead with merging this.

**Comment by aravind-pg** [comment]
Yeah I'm a bit surprised by it too. Will try to look into it more and come back if I have questions.

**Comment by aravind-pg** [comment]
@jdm Sorry for the delay here, I've been a bit busy with things and will continue to be for a few days more. If you have any suggestions (just high-level stuff off the top of your head) for where to start poking around to see why the tests are failing that would be super useful, but no problem at all if not, I'll just poke around myself (in a few days).

**Comment by jdm** [comment]
Aha!
1. we removed generic/subresources/worker.py
2. we need to reinstate it with a change like https://dxr.mozilla.org/servo/source/tests/wpt/mozilla/tests/mozilla/referrer-policy/generic/subresource/document.py#4
That should give us much more meaningful results.

**Comment by aravind-pg** [comment]
@jdm Really sorry for vanishing for so long, was a busy time for me.

So I reinstated worker.py as per your suggestion, but unfortunately it seems not to have made any difference. I'm really not sure what's going on here, but I'll try to dig some more.

**Comment by jdm** [comment]
Does worker.js.template exist? If not we need to reinstate it as well.

**Comment by aravind-pg** [comment]
Does look like it does: [worker.js.template](https://github.com/aravind-pg/servo/blob/b423496583e253e0de07dc61eaaa016552c933ec/tests/wpt/mozilla/tests/mozilla/referrer-policy/generic/template/worker.js.template).

**Comment by jdm** [comment]
I took a look at the branch locally and figured out what's going on. b423496 isn't necessary any more since we're enabling the worker tests in the upstream tests, rather than our local fork. The reason why the tests are failing is that they are testing the reported referrer for the initial worker object that is created, while the changes in 33007c2 and 53cdc62 are in support of referrers for requests created _from_ a worker script. Fixing the tests should only require modifiying the `WorkerScriptLoadOrigin` object in `Worker::Constructor` to use the document's referrer policy; however, we should add a new set of tests that initiates requests via `XMLHttpRequest` inside of a worker to ensure that the changes in this PR actually work.

**Comment by aravind-pg** [comment]
Very sorry for the extend to which this stagnated, @jdm. I think I'll just close this PR for now and perhaps open a new one when I find some time. Thanks a lot for your help, and I'll keep your last comment in mind when I reopen.

**Pull request closed without merging**